File: cli/src/sync/AggregateTypes.ts
/**
 * Type-only definitions for the four `.jolli/<aggregate>.json` files and the
 * content-addressed per-commit summary files (JOLLI-1316 §2 schema).
 *
 * The backend's `JolliMemoryAggregateValidator` enforces well-formedness on
 * push; client-side `AggregateMerge` produces values that pass validation.
 * Both ends ship independent implementations of the same schema — keep them
 * in lockstep (see `jolli-1316-aggregate-merge-design.md §6 + §7`).
 *
 * Pure types — no runtime imports — so every consumer (CLI, plugin, future
 * tooling) imports from one source without dragging in a runtime module.
 */

// ── manifest.json ─────────────────────────────────────────────────────────

/**
 * `manifest.json` entry — one row per content file. `fileId` is the primary
 * key; `source.generatedAt` is the merge tiebreaker (§3.1).
 */
export interface ManifestEntry {
	readonly path: string;
	readonly fileId: string;
	readonly type: "commit";
	readonly fingerprint: string;
	readonly title: string;
	readonly source: {
		readonly commitHash: string;
		readonly branch: string;
		readonly generatedAt: string; // ISO8601
	};
}

export interface ManifestEnvelope {
	readonly version: 1;
	readonly files: ReadonlyArray<ManifestEntry>;
}

// ── index.json ────────────────────────────────────────────────────────────

/**
 * `index.json` entry — one row per commit (or amend). Primary key
 * `commitHash`; merge tiebreak uses the 2×2 of `parentCommitHash` × `generatedAt`
 * (§3.2 — null-parent rows always lose to ones with a parent).
 */
export interface IndexEntry {
	readonly commitHash: string;
	readonly parentCommitHash: string | null;
	readonly treeHash: string;
	readonly commitType: "commit" | "amend";
	readonly commitMessage: string;
	readonly commitDate: string; // ISO8601
	readonly branch: string;
	readonly generatedAt: string; // ISO8601
	readonly topicCount?: number;
	readonly diffStats?: {
		readonly filesChanged: number;
		readonly insertions: number;
		readonly deletions: number;
	};
}

export interface IndexEnvelope {
	readonly version: 3;
	readonly entries: ReadonlyArray<IndexEntry>;
}

// ── branches.json ─────────────────────────────────────────────────────────

/**
 * `branches.json` entry — one row per branch → folder mapping. Primary key
 * `branch`; no tiebreak (last-write-wins is fine because the
 * `folder === canonical(branch)` invariant is enforced at write time, so any
 * two valid entries for the same branch are identical save for `createdAt`).
 */
export interface BranchEntry {
	readonly folder: string;
	readonly branch: string;
	readonly createdAt: string; // ISO8601
}

export interface BranchesEnvelope {
	readonly version: 1;
	readonly mappings: ReadonlyArray<BranchEntry>;
}

// ── catalog.json ──────────────────────────────────────────────────────────

/**
 * `catalog.json` entry — long-form recap + topic breakdown per commit.
 * Primary key `commitHash`; no tiebreak (§3.4 — both sides produced the
 * same content for a given commit, so identity suffices).
 */
export interface CatalogTopic {
	readonly title: string;
	readonly decisions: string;
	readonly category: string;
	readonly importance: string;
	readonly filesAffected: ReadonlyArray<string>;
}

export interface CatalogEntry {
	readonly commitHash: string;
	readonly recap: string;
	readonly ticketId: string;
	readonly topics: ReadonlyArray<CatalogTopic>;
}

export interface CatalogEnvelope {
	readonly version: 1;
	readonly entries: ReadonlyArray<CatalogEntry>;
}

// ── summaries/<commitHash>.json ───────────────────────────────────────────

/**
 * Content-addressed per-commit summary. Filename hash MUST equal the embedded
 * `commitHash` (backend validator rejects mismatches). No merge — git default
 * is fine because identical filenames carry identical content.
 *
 * The plugin doesn't dictate the body shape beyond `commitHash`; downstream
 * consumers (web UI, analytics) interpret remaining fields. Keep this loose
 * so summary generators on different branches don't have to agree.
 */
export interface SummaryFile {
	readonly commitHash: string;
	readonly [k: string]: unknown;
}

File: intellij/src/main/kotlin/ai/jolli/jollimemory/sync/AggregateTypes.kt
package ai.jolli.jollimemory.sync

/**
 * Type definitions for the four `.jolli/<aggregate>.json` files and the
 * content-addressed per-commit summary files (JOLLI-1316 §2 schema).
 *
 * Port of `cli/src/sync/AggregateTypes.ts`.
 *
 * Pure data — no runtime logic — so every consumer imports from one source.
 */

// ── manifest.json ──────────────────────────────────────────────────────

data class ManifestSource(
	val commitHash: String,
	val branch: String,
	val generatedAt: String, // ISO8601
)

data class ManifestEntry(
	val path: String,
	val fileId: String,
	val type: String, // "commit"
	val fingerprint: String,
	val title: String,
	val source: ManifestSource,
)

data class ManifestEnvelope(
	val version: Int, // 1
	val files: List<ManifestEntry>,
)

// ── index.json ─────────────────────────────────────────────────────────

data class DiffStats(
	val filesChanged: Int,
	val insertions: Int,
	val deletions: Int,
)

data class IndexEntry(
	val commitHash: String,
	val parentCommitHash: String?, // null when no parent
	val treeHash: String,
	val commitType: String, // "commit" | "amend"
	val commitMessage: String,
	val commitDate: String, // ISO8601
	val branch: String,
	val generatedAt: String, // ISO8601
	val topicCount: Int? = null,
	val diffStats: DiffStats? = null,
)

data class IndexEnvelope(
	val version: Int, // 3
	val entries: List<IndexEntry>,
)

// ── branches.json ──────────────────────────────────────────────────────

data class BranchEntry(
	val folder: String,
	val branch: String,
	val createdAt: String, // ISO8601
)

data class BranchesEnvelope(
	val version: Int, // 1
	val mappings: List<BranchEntry>,
)

// ── catalog.json ───────────────────────────────────────────────────────

data class CatalogTopic(
	val title: String,
	val decisions: String,
	val category: String,
	val importance: String,
	val filesAffected: List<String>,
)

data class CatalogEntry(
	val commitHash: String,
	val recap: String,
	val ticketId: String,
	val topics: List<CatalogTopic>,
)

data class CatalogEnvelope(
	val version: Int, // 1
	val entries: List<CatalogEntry>,
)
